File: api/models/order.js
const mongoose = require("mongoose");

const OrderProductSchema = new mongoose.Schema(
  {
    productId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
      required: true,
    },
    name: {
      type: String,
      required: true,
      trim: true,
    },
    price: {
      type: Number,
      required: true,
      min: 0,
    },
    quantity: {
      type: Number,
      required: true,
      min: 1,
    },
    allowSubstitution: {
      type: Boolean,
      default: false,
    },
    substitutedWith: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
      default: null,
    },
    substitutedName: {
      type: String,
      default: null,
    },
    originalPrice: {
      type: Number,
      default: null,
    },
  },
  { _id: false }
);

const StockAllocationSchema = new mongoose.Schema(
  {
    lotId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true,
    },
    quantity: {
      type: Number,
      required: true,
      min: 1,
    },
  },
  { _id: false }
);

const OrderAllocationSchema = new mongoose.Schema(
  {
    productId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Product",
      required: true,
    },
    lotAllocations: {
      type: [StockAllocationSchema],
      default: [],
    },
  },
  { _id: false }
);

const OrderSchema = new mongoose.Schema(
  {
    orderNumber: {
      type: String,
      unique: true,
      required: true,
      trim: true,
    },
    userId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "User",
      required: [true, "L'utilisateur est requis"],
    },
    clientId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Client",
      required: [true, "Le client est requis"],
    },
    companyId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Company",
      required: [true, "La société est requise"],
    },
    products: {
      type: [OrderProductSchema],
      default: [],
      validate: {
        validator: (products) => Array.isArray(products) && products.length > 0,
        message: "La commande doit contenir au moins un article",
      },
    },
    totalPrice: {
      type: Number,
      required: true,
      min: 0,
      default: 0,
    },
    status: {
      type: String,
      enum: ["en_attente", "en_preparation", "pret_pour_recuperation", "validee", "annulee"],
      default: "en_attente",
    },

    stockAllocations: {
      type: [OrderAllocationSchema],
      default: [],
    },
    notes: {
      type: String,
      trim: true,
      maxlength: 500,
    },
    pickupMode: {
      type: String,
      enum: ["sur_place", "livraison"],
      default: "sur_place",
    },
    invoiceNumber: {
      type: String,
      unique: true,
      sparse: true,
    },
    invoiceDate: Date,
    collectionCode: {
      type: String,
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

OrderSchema.index({ companyId: 1, createdAt: -1 });
OrderSchema.index({ clientId: 1, createdAt: -1 });
OrderSchema.index({ userId: 1, createdAt: -1 });
OrderSchema.index({ status: 1 });

module.exports = mongoose.model("Order", OrderSchema);

File: client_app/lib/models/order.dart
String orderStatusLabel(String status) {
  switch (status) {
    case 'en_attente':
      return 'En attente';
    case 'validee':
      return 'Validée';
    case 'en_preparation':
      return 'En préparation';
    case 'pret_pour_recuperation':
      return 'Prête pour récupération';
    case 'annulee':
      return 'Annulée';
    default:
      return status;
  }
}

class OrderItem {
  final String productId;
  final String name;
  final double price;
  final int quantity;
  final bool allowSubstitution;
  final String? substitutedWith;
  final String? substitutedName;
  final double? originalPrice;

  const OrderItem({
    required this.productId,
    required this.name,
    required this.price,
    required this.quantity,
    this.allowSubstitution = false,
    this.substitutedWith,
    this.substitutedName,
    this.originalPrice,
  });

  double get subtotal => price * quantity;
  bool get wasSubstituted => substitutedWith != null;

  factory OrderItem.fromJson(Map<String, dynamic> json) {
    final product = json['productId'] ?? json['product'];
    final resolvedProductId = product is Map<String, dynamic>
        ? (product['_id'] ?? '').toString()
        : (product ?? '').toString();

    return OrderItem(
      productId: resolvedProductId,
      name: (json['name'] ?? '').toString(),
      price: ((json['price'] ?? 0) as num).toDouble(),
      quantity: ((json['quantity'] ?? 0) as num).toInt(),
      allowSubstitution: json['allowSubstitution'] == true,
      substitutedWith: json['substitutedWith']?.toString(),
      substitutedName: json['substitutedName']?.toString(),
      originalPrice: json['originalPrice'] != null
          ? (json['originalPrice'] as num).toDouble()
          : null,
    );
  }

  OrderItem copyWith({bool? allowSubstitution}) {
    return OrderItem(
      productId: productId,
      name: name,
      price: price,
      quantity: quantity,
      allowSubstitution: allowSubstitution ?? this.allowSubstitution,
      substitutedWith: substitutedWith,
      substitutedName: substitutedName,
      originalPrice: originalPrice,
    );
  }

  Map<String, dynamic> toRequestJson() {
    return {
      'productId': productId,
      'quantity': quantity,
      'allowSubstitution': allowSubstitution,
    };
  }
}

class Order {
  final String id;
  final String orderNumber;
  final String userId;
  final String clientId;
  final List<OrderItem> items;
  final double totalPrice;
  final String status;
  final String? notes;
  final DateTime createdAt;
  final DateTime updatedAt;
  final String? invoiceNumber;
  final DateTime? invoiceDate;
  final String? collectionCode;
  final String pickupMode;

  const Order({
    required this.id,
    required this.orderNumber,
    required this.userId,
    required this.clientId,
    required this.items,
    required this.totalPrice,
    required this.status,
    this.notes,
    required this.createdAt,
    required this.updatedAt,
    this.invoiceNumber,
    this.invoiceDate,
    this.collectionCode,
    this.pickupMode = 'sur_place',
  });

  String get statusLabel => orderStatusLabel(status);
  bool get isTerminal => status == 'validee' || status == 'annulee';
  bool get hasInvoice => (invoiceNumber ?? '').trim().isNotEmpty;
  bool get hasSubstitutions => items.any((i) => i.wasSubstituted);

  factory Order.fromJson(Map<String, dynamic> json) {
    final user = json['userId'];
    final client = json['clientId'];
    final products =
        (json['products'] as List<dynamic>?) ??
        (json['items'] as List<dynamic>?) ??
        [];

    return Order(
      id: (json['_id'] ?? '').toString(),
      orderNumber: (json['orderNumber'] ?? '').toString(),
      userId: user is Map<String, dynamic>
          ? (user['_id'] ?? '').toString()
          : (user ?? '').toString(),
      clientId: client is Map<String, dynamic>
          ? (client['_id'] ?? '').toString()
          : (client ?? '').toString(),
      items: products
          .map((item) => OrderItem.fromJson(item as Map<String, dynamic>))
          .toList(),
      totalPrice: ((json['totalPrice'] ?? json['total'] ?? 0) as num)
          .toDouble(),
      status: (json['status'] ?? 'en_attente').toString(),
      notes: json['notes']?.toString(),
      createdAt:
          DateTime.tryParse((json['createdAt'] ?? '').toString()) ??
          DateTime.now(),
      updatedAt:
          DateTime.tryParse((json['updatedAt'] ?? '').toString()) ??
          DateTime.now(),
      invoiceNumber: json['invoiceNumber']?.toString(),
      invoiceDate: json['invoiceDate'] != null
          ? DateTime.tryParse(json['invoiceDate'].toString())
          : null,
      collectionCode: json['collectionCode']?.toString(),
      pickupMode: (json['pickupMode'] ?? 'sur_place').toString(),
    );
  }
}
